core/src/blocktree.rs

```diff
@@ -921,6 +966,41 @@ impl Blocktree {
         batch_processor.write(batch)?;
         Ok(())
     }
+
+    /// Prune blocktree such that slots higher than `target_slot` are deleted and all references to
+    /// higher slots are removed
+    pub fn prune(&self, target_slot: u64) {
+        let mut meta = self
+            .meta(target_slot)
+            .expect("couldn't read slot meta")
+            .expect("no meta for root slot");
+        meta.next_slots.clear();
+        self.put_meta_bytes(
+            target_slot,
+            &bincode::serialize(&meta).expect("couldn't get meta bytes"),
+        )
+        .expect("unable to update meta");
+
+        self.slot_meta_iterator(target_slot + 1)
+            .expect("unable to iterate over meta")
+            .for_each(|(slot, _)| self.delete_slot(slot).unwrap());
+
+        // fixup anything that refers to non-root slots and delete the rest
+        for (slot, mut meta) in self
+            .slot_meta_iterator(0)
+            .expect("unable to iterate over meta")
+        {
+            meta.next_slots.retain(|slot| *slot <= target_slot);
+            self.put_meta_bytes(
+                slot,
+                &bincode::serialize(&meta).expect("couldn't update meta"),
+            )
+            .expect("couldn't update meta");
+            if slot > target_slot {
+                break;
+            }
+        }
+    }
```

Comment:
consistency check at the end of this function?

Reply:
or maybe another function that looks like fsck() that a tool can call

Reply:
filed this as a separate issue https://github.com/solana-labs/solana/issues/5156